plugins/cros-ec/fu-cros-ec-firmware.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2020 Benson Leung <bleung@chromium.org>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-fmap-firmware.h"
+#include "fu-cros-ec-common.h"
+#include "fu-cros-ec-firmware.h"
+
+#define MAXSECTIONS	2
+
+struct _FuCrosEcFirmware {
+	FuFirmware			parent_instance;
+	struct cros_ec_version		version;
+	FuCrosEcFirmwareSection		sections[MAXSECTIONS];
+};
+
+G_DEFINE_TYPE (FuCrosEcFirmware, fu_cros_ec_firmware, FU_TYPE_FIRMWARE)
+
+static gboolean
+fu_cros_ec_firmware_parse (FuFirmware *firmware,
+			   GBytes *fw,
+			   guint64 addr_start,
+			   guint64 addr_end,
+			   FwupdInstallFlags flags,
+			   GError **error)
+{
+	FuCrosEcFirmware *self = FU_CROS_EC_FIRMWARE (firmware);
+	g_autoptr(FuFirmware) fmap_firmware = fu_fmap_firmware_new ();
+
+	self->sections[0].name = "RO";
+	self->sections[1].name = "RW";
+
+	if (!fu_firmware_parse (fmap_firmware, fw, flags, error))
+		return FALSE;
+
+	for (gsize i = 0; i < G_N_ELEMENTS (self->sections); i++) {
+		gboolean rw = FALSE;
+		FuCrosEcFirmwareSection section = self->sections[i];
+		const gchar *fmap_name;
+		const gchar *fmap_fwid_name;
+		g_autoptr(FuFirmwareImage) img = NULL;
+		g_autoptr(FuFirmwareImage) fwid_img = NULL;
+		g_autoptr(GBytes) payload_bytes = NULL;
+		g_autoptr(GBytes) fwid_bytes = NULL;
+
+		if (g_strcmp0 (section.name, "RO") == 0) {
+			fmap_name = "EC_RO";
+			fmap_fwid_name = "RO_FRID";
+		} else if (g_strcmp0 (section.name, "RW") == 0) {
+			rw = TRUE;
+			fmap_name = "EC_RW";
+			fmap_fwid_name = "RW_FWID";
+		} else {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_INVALID_DATA,
+					     "incorrect section name");
+			return FALSE;
+		}
+
+		img = fu_firmware_get_image_by_id (fmap_firmware,
+						   fmap_name, error);
+		if (img == NULL) {
+			g_prefix_error (error, "%s image not found: ",
+					fmap_name);
+			return FALSE;
+		}
+
+		fwid_img = fu_firmware_get_image_by_id (fmap_firmware,
+							fmap_fwid_name, error);
+		if (fwid_img == NULL) {
+			g_prefix_error (error, "%s image not found: ",
+					fmap_fwid_name);
+			return FALSE;
+		}
+		fwid_bytes = fu_firmware_image_write(fwid_img, error);
+		if (fwid_bytes == NULL) {
+			g_prefix_error (error,
+					"unable to get bytes from %s: ",
+					fmap_fwid_name);
+			return FALSE;
+		}
+		memcpy (section.version,
+			g_bytes_get_data (fwid_bytes, NULL),
+			g_bytes_get_size (fwid_bytes));
```

Comment:
this makes me a bit twitchy. Maybe use `fu_memcpy_safe`?

Reply:
Done. Also added another change to replace the memcpy in fu-cros-ec-usb-device.c as well.